---
title: "BMI511 Singe Cell Breast Cancer Dataset Analysis"
author: "Erica Brophy"
date: "5/4/2020"
output:
  html_document: default
  pdf_document: default
---

```
Single cell RNA-seq of circulating breast cancer cells from 19 women with 
ER+/HER2- primary tumours. Analysis of 74 single candidate circultaing tumor 
cells (CTCs) from two representative ER+/HER2- patients (BRx-42 and BRx-82) 
and 14 triple negative patients were isolated via micromanipulation and 
subjected to single-cell RNA-sequencing (other patient samples were dropped 
for low quality). The goal is to distinguish a subpopulation of HER2+ cells from 
primarily HER2- CTCs that forms following multiple courses of therapy. 
The method used for analysis was Seurat, an R toolkit for single cell genomics.

Citations:
Jordan, Nicole Vincent et al. “HER2 expression identifies dynamic functional states within circulating breast cancer cells.” Nature vol. 537,7618 (2016): 102-106. 

Butler, A., Hoffman, P., Smibert, P. et al. "Integrating single-cell transcriptomic data across different conditions, technologies, and species." Nat Biotechnol 36, 411–420 (2018).

```

### Loading dataset
```{r echo=TRUE, message=FALSE, warning=FALSE}
library(Seurat)
library(dplyr)
library(patchwork)
library(Matrix)
library(readr)
library(clusterProfiler)
library(org.Hs.eg.db)
library(plyr)


data <- readMM("/Users/ericabrophy/Documents/BMI511_translational/BreastCancerDataset/E-GEOD-75367.aggregated_filtered_counts.mtx")
data_col <- read_delim("/Users/ericabrophy/Documents/BMI511_translational/BreastCancerDataset/E-GEOD-75367.aggregated_filtered_counts.mtx_cols", 
                       "\t", escape_double = FALSE, col_names = FALSE, 
                       trim_ws = TRUE)
data_row <- read_delim("/Users/ericabrophy/Documents/BMI511_translational/BreastCancerDataset/E-GEOD-75367.aggregated_filtered_counts.mtx_rows", 
                           "\t", escape_double = FALSE, col_names = FALSE, 
                            trim_ws = TRUE)
metadata <- read_delim("/Users/ericabrophy/Documents/BMI511_translational/BreastCancerDataset/E-GEOD-75367.sdrf.txt", 
                       "\t", escape_double = FALSE, trim_ws = TRUE)
```


### Reshaping data
```{r}
data <- as.matrix(data)
data_row <- data_row[, -1]

rownames(data)<-data_row$X2
colnames(data)<-data_col$X1

data.temp = data
data.temp[data.temp==0] = NA
hist(log10(as.vector(data.temp)+ 1))
```

### Assign actual gene names to IDs 
```{r message=FALSE, warning=FALSE}
 gene.df <- bitr(data_row$X2, fromType = "ENSEMBL",
                 toType = c("ENSEMBL", "SYMBOL"),
                 OrgDb = org.Hs.eg.db)

```

### Pull out translated gene names 
```{r message=FALSE, warning=FALSE}
gene_symbol <- gene.df[,2]
```

### Summary of the data
```{r message=FALSE, warning=FALSE}
head(gene.df, n=5)
summary(gene.df, n=5)
```

### Subsetting new data
```{r results= "hide"}
intersection <- gene.df[,1] %in% rownames(data) 
data[intersection, ]
data_subset <- data[gene.df[,1], ]
data_subset <- as.data.frame(data_subset)
```

### Match gene labels for both dataframes
```{r echo=TRUE, message=FALSE, warning=FALSE}
genomic_idx <- match(rownames(data_subset), gene.df[,1])
```

### Renaming rows of new data matrix to gene names
```{r message=FALSE, warning=FALSE}
rownames(data_subset) = make.names(gene.df$SYMBOL, unique=TRUE)
```

### The Seurat object serves as a container that contains both data and analysis
```{r message=FALSE, warning=FALSE}
BCdata <- CreateSeuratObject(counts = data_subset, 
                             project = "BreastCancer", 
                             min.cells = 3, 
                             min.features = 200)
FeatureScatter(BCdata,"nCount_RNA","nFeature_RNA")

```

### Check for MT genes -> QC
```{r results= "hide"}
rownames(data_subset) = make.names(gene.df$SYMBOL, unique=TRUE)
head(grep(pattern = "^MT-", x = rownames(data_subset), value = TRUE))
BCdata[grep(pattern = "^MT-", x = rownames(data_subset))]
```

### QC
##### Look at mitochondrial transcript levels to filter out apoptotic, stressed, and low-quality cells.
```{r warning=FALSE}
BCdata[["percent.mt"]] <- PercentageFeatureSet(BCdata, pattern = "^MT-")
head(BCdata@meta.data, 5) #There are no MT genes in the dataset
VlnPlot(BCdata, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

```


### Normalize data
##### Normalizes the feature expression measurements for each cell by the total expression.
```{r results= "hide"}
#multiplies this by a scale factor (10,000 by default), and log-transforms the result
BCdata <- NormalizeData(BCdata, normalization.method = "LogNormalize", scale.factor = 10000)
BCdata[["RNA"]]@data
```

### Barplot of features across dataset
```{r}
barplot(BCdata@meta.data$nFeature_RNA)
```

### Feature Selection
```{r message=FALSE, warning=FALSE, results= "hide"}
#A subset of features that exhibit high cell-to-cell variation in the dataset 
BCdata <- FindVariableFeatures(BCdata, selection.method = "vst", nfeatures = 3000)
```

### Plot variable features with and without labels 
##### Plot features that are expressing cell to cell varitation in the data to identify genes that may be highly expressed in some cells and lowly expressed in others.
```{r message=FALSE, warning=FALSE}
# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(BCdata), 10)
LabelPoints(plot = VariableFeaturePlot(BCdata), points = top10, repel = TRUE)
```

### Scaling the data
##### This gives equal weight to all genes in downstream analysis so that high expressing genes do not dominate. 
```{r echo=TRUE, message=FALSE, warning=FALSE, results="hide"}
#apply a linear transformation (‘scaling’) that is a standard pre-processing 
#step prior to dimensional reduction techniques like PCA. 
all.genes <- rownames(BCdata)
BCdata <- ScaleData(BCdata, features = all.genes)
BCdata[["RNA"]]@scale.data
```

### PCA
##### Dimension reduction to visualize patterns in the data correlating to high variance.
```{r message=FALSE, warning=FALSE}
BCdata <- RunPCA(BCdata, features = VariableFeatures(object = BCdata), npcs = 15, verbose = FALSE, approx=FALSE)
print(BCdata[["pca"]], dims = 1:5, nfeatures = 5)
VizDimLoadings(BCdata, dims = 1:2, reduction = "pca")
DimPlot(BCdata, reduction = "pca")
```

### DimHeatmap
##### Cells and features are ordered according to their PCA scores. This allows you to explore heterogeneity in the dataset.
```{r message=FALSE, warning=FALSE}
#Easy exploration of the primary sources of heterogeneity in a dataset
#and can be useful when trying to decide which PCs to include for further downstream analyses
DimHeatmap(BCdata, dims = 1, cells = 74, balanced = TRUE)
DimHeatmap(BCdata, dims = 1:5, cells = 74, balanced = TRUE)
```

### Jackstraw Plot
##### Visualization tool for comparing the distribution of p-values for each PC with a uniform distribution. Elbow plot exhibits a ranking of principle components based on the percentage of variance explained by each one.
```{r message=FALSE, warning=FALSE}
#a ranking of principle components based on the percentage of variance explained by each one
BCdata <- JackStraw(BCdata, num.replicate = 100)
BCdata <- ScoreJackStraw(BCdata, dims = 1:10)
JackStrawPlot(BCdata, dims = 1:10)
ElbowPlot(BCdata)
```

### Clustering cells
##### Clustering methodsembed cells in a graph structure. 
```{r message=FALSE, warning=FALSE}
BCdata <- FindNeighbors(BCdata, dims = 1:10)
BCdata <- FindClusters(BCdata, resolution = 0.5)
head(Idents(BCdata), 5)
```

### UMAP
##### Non-linear dimensional reduction techniques to visualize and explore the dataset.
```{r message=FALSE, warning=FALSE}
BCdata <- RunUMAP(BCdata, dims = 1:15)
DimPlot(BCdata, reduction = "umap")
```

### Find all markers for a cluster -> cluster 1
##### Can find markers that define clusters via differential expression.
```{r message=FALSE, warning=FALSE}
cluster1.markers <- FindMarkers(BCdata, ident.1 = 1, min.pct = 0.25)
head(cluster1.markers, n = 5)
```

### Find markers for every cluster compared to all remaining cells, report only the positive ones
```{r message=FALSE, warning=FALSE}
BCdata.markers <- FindAllMarkers(BCdata, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
BCdata.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC)
```


### The ROC test returns the ‘classification power’ for any individual marker (ranging from 0 - random, to 1 - perfect)
```{r message=FALSE, warning=FALSE}
cluster1.markers <- FindMarkers(BCdata, ident.1 = 0, logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
head(cluster1.markers, n = 5)
```

### Shows visualization of marker expression and marker expression probability distributions across clusters
```{r message=FALSE, warning=FALSE}
VlnPlot(BCdata, features = c("STX12", "HBA1"))
FeaturePlot(BCdata, features = c("STX12", "ATL3", "PTPA", "YIPF5", "MARCHF5", "HBD", "HBA1", "HBA2","HBB", "CA1"))
```

### Generates an expression heatmap for given cells and features for top 10 highly expressing genes
```{r message=FALSE, warning=FALSE}

top10 <- BCdata.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
DoHeatmap(BCdata, features = top10$gene) + NoLegend()

```

### Naming clusters
```{r message=FALSE, warning=FALSE}

new.cluster.ids <- c("HER2-", "HER2+")
names(new.cluster.ids) <- levels(BCdata)
BCdata <- RenameIdents(BCdata, new.cluster.ids)
plot <- DimPlot(BCdata, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
HoverLocator(plot = plot, information = FetchData(BCdata, vars = c("ident", "PC_1", "nFeature_RNA")))
```

### TSNE
##### Non-linear dimensional reduction techniques to visualize and explore the dataset.
```{r message=FALSE, warning=FALSE}

BCdata <- RunTSNE(object = BCdata, dims = 1:3, resolution = 0.5, perplexity = 15)
plot <- DimPlot(object = BCdata, reduction = "tsne", label = TRUE)
HoverLocator(plot = plot, information = FetchData(BCdata, vars = c("ident", "PC_1", "nFeature_RNA")))
```

### Dotplot of genes identified showing expression levels in the different clusters 
```{r message=FALSE, warning=FALSE}
features <- c("STX12", "ATL3", "PTPA", "YIPF5", "MARCHF5", "HBD", "HBA1", "HBA2","HBB", "CA1")
features2 <- c("APOD", "PIP", "CRABP2", "XBP1", "HSP90AB1", "F13A1", "TUBB1", "RGS18", "H2AC6", "CAVIN2", "CD226")
DotPlot(BCdata, features = features) + RotatedAxis()
DotPlot(BCdata, features = features2) + RotatedAxis()
```

### Heatmap showing expression levels of genes identified in clusters 
```{r message=FALSE, warning=FALSE}
DoHeatmap(subset(BCdata, downsample = 100), features = features2, size = 3)
```

### Interactive Plot to show ERBB2 expression in data
```{r message=FALSE, warning=FALSE}
plot <- FeaturePlot(BCdata, features = "ERBB2")
HoverLocator(plot = plot, information = FetchData(BCdata, vars = c("ident", "PC_1", "nFeature_RNA")))
```


### PCA plot with labels
```{r message=FALSE, warning=FALSE}
plot <- DimPlot(BCdata, reduction = "pca", label = TRUE)
HoverLocator(plot = plot, information = FetchData(BCdata, vars = c("ident", "PC_1", "nFeature_RNA")))
```

### Shiny App URL 
```{r}
#https://ebrophy.shinyapps.io/SingleCellBreastCancer_Heatmap/
```


